kbfsblock/id.go

```diff
@@ -91,6 +104,37 @@ func MakeTemporaryID() (ID, error) {
 	return ID{h}, nil
 }
 
+// MakeRandomIDInRange generates a random block ID using a CSPRNG, distributing
+// the random variable over the interval [start, end), where the full range is
+// [0, MaxUint64). This corresponds to a normalized representation of the
+// range [kbfshash.RawDefaultHash{}, kbfshash.MaxDefaultHash).
+func MakeRandomIDInRange(start, end uint64) (ID, error) {
+	if start >= end {
+		return ID{}, errors.New("Invalid range for random ID")
+	}
+	rangeSize := float64(end - start)
+	randBuf := make([]byte, 8)
+	err := kbfscrypto.RandRead(randBuf)
+	if err != nil {
+		return ID{}, err
+	}
+	// Generate a random unsigned int. Endianness doesn't matter here because
+	// the bytes are random.
+	randUint := binary.BigEndian.Uint64(randBuf)
+	randFloat := float64(randUint) / float64(math.MaxUint64)
```

Comment:
The distribution of floats between 0 and 1 is logarithmic. The best way I thought of doing a true uniform distribution is generating only 6 bytes (mantissa is 53 bits, 6 bytes is 48 bits, close enough), using that with a fixed exponent (1), and then doing all the math with that. As long as you divide by a power of 2 (2^48 in this case), you're only changing the exponent and thus have no loss of precision. I don't think the conversions are a big deal given that this operation is never performed on a tight loop, and there are far greater bottlenecks on the critical path (writing to leveldb).